src/core/atomic.d

```diff
@@ -1273,6 +1182,146 @@ else version( AsmX86_64 )
     }
 }
 
+/* ***
+ * Functions built in to the compiler, along with a default implementation.
+ */
+
+bool cmpxchg(shared(ubyte)* here, ubyte ifThis, ubyte writeThis) pure nothrow @nogc @trusted
```

Comment:
Also we're going to need a lot more intrinsics that that, https://gcc.gnu.org/onlinedocs/gcc-4.9.3/gcc/_005f_005fatomic-Builtins.html. You don't want to add all of those, particularly considering that they are dependent on the compiler.